site/static/api/device-registry-v1.yaml

```diff
@@ -0,0 +1,745 @@
+               $ref: '#/components/responses/NotFound'
+
+      put:
+         tags:
+            - devices
+         summary: Update existing device registration
+         operationId: updateRegistration
+         parameters:
+            - $ref: '#/components/parameters/resourceVersion'
+         requestBody:
+            description: Updated device registration
+            content:
+               application/json:
+                  schema:
+                     $ref: '#/components/schemas/Device'
+            required: true
+         responses:
+            204:
+               $ref: '#/components/responses/Updated'
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            403:
+               $ref: '#/components/responses/NotAllowed'
+            404:
+               $ref: '#/components/responses/NotFound'
+            412:
+               $ref: '#/components/responses/ResourceVersionMismatch'
+
+      delete:
+         tags:
+            - devices
+         summary: Delete device registration
+         operationId: deleteRegistration
+         parameters:
+            - $ref: '#/components/parameters/resourceVersion'
+         responses:
+            204:
+               $ref: '#/components/responses/Delete'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            403:
+               $ref: '#/components/responses/NotAllowed'
+            404:
+               $ref: '#/components/responses/NotFound'
+            412:
+               $ref: '#/components/responses/ResourceVersionMismatch'
+
+   /credentials/{tenantId}/{deviceId}:
+
+      parameters:
+         - $ref: '#/components/parameters/tenantId'
+         - $ref: '#/components/parameters/deviceId'
+
+      get:
+         tags:
+            - credentials
+         summary: Get credentials set of a device.
+         description: |
+            Get the credentials set of a device. As long as the device is
+            registered and the user has read access to it, this call should
+            never return "not found".
+         operationId: getAllCredentials
+         responses:
+            200:
+               description: Operation successful
+               content:
+                  application/json:
+                     schema:
+                        $ref: '#/components/schemas/CredentialsSet'
+               headers:
+                  ETag:
+                     description: Version of the resource
+                     schema:
+                        type: string
+            400:
+               $ref: '#/components/responses/MalformedRequest'
+            401:
+               $ref: '#/components/responses/Unauthorized'
+            404:
+               $ref: '#/components/responses/NotFound'
+
+      put:
+         tags:
+            - credentials
+         summary: Update credentials set for registered device
```

Comment:
The only way to add a credentials seems to be getting all credentials, then modifying, deleting or adding an entry and then putting the result back. This requires some client side logic  for JSON parsing to find the right entry. Dedicated URLS for putting and deleting an entry would add much more convenience for the client. A use case like credential rotation can then be implemented without any parsing.

PUT /credentials/{tenantId}/{deviceId}/{auth-id}
DELETE /credentials/{tenantId}/{deviceId}/{auth-id}

Updating a set at once (PUT /credentials/{tenantId}/{deviceId}}) may not be needed when individual URLs are provided.
Deleting all at once would still be convenient for cleaning up all credentials at once. 

Was the option of providing dedicated auth id update considered? 
This is just an API consumers view, there might be some reasons internally, why this isn't a good idea :-)

Reply:
makes sense to me. However, the resources would need to include the credentials type as well, i.e.
```
PUT /credentials/{tenantId}/{deviceId}/{type}/{auth-id}
DELETE /credentials/{tenantId}/{deviceId}/{type}/{auth-id}
```

Reply:
> Was the option of providing dedicated auth id update considered?

> However, the resources would need to include the credentials type

That is how the current API look like : https://current-api-hono-device-registry-api.wonderful.iot-playground.org/
I am not sure why those operation have been removed. Also : how does one create a credential if none exist yet ?

Reply:
I touched that point in the following comment: https://github.com/eclipse/hono/pull/1130#issuecomment-487902225

I agree, I think it makes sense to have such specialized operations. However they can all be implemented on top of this API. And I think it makes sense to add them in a later version of the API. However I don't think it conflicts with the current API definition.

Assuming you have all kinds of specialized secrets (Password, PSK, client cert, custom protocol adapter secrets) I think it makes sense to support them in the way currently described. And then adding more specific operations on top of it, like:

* Cleanup expired passwords
* Password / Key rollover
* …

So, let's finish this first. And then do version 1.1, including search and specialized secret operations.